symengine/sets.cpp

```diff
@@ -0,0 +1,223 @@
+				}
+				else {
+					start = this->start_;
+					left_open = this->left_open_;
+				}
+			}
+			else {
+				start = this->start_;
+				left_open = this->left_open_ or other.left_open_ ;
+			}
+
+			if (neq(*this->end_, *other.end_)) {
+				if (eq(*this->end_, *end_end)) {
+					end = this->end_;
+					right_open = this->right_open_;
+				}
+				else {
+					end = other.end_;
+					right_open = other.right_open_;
+				}
+			}
+			else {
+				end = this->end_;
+				right_open = this->right_open_ or other.right_open_;
+			}
+			return interval(start, end, left_open, right_open);
+		}
+		else {
+			return emptyset();
+		}
+	}
+
+	RCP<const Set> Interval::_union(const Set &o) const {
+		SYMENGINE_ASSERT(is_a<Interval>(o));
+		const Interval &other = static_cast<const Interval &>(o);
+		RCP<const Basic> start_start, end_end, m;
+		RCP<const Number> start, end;
+		bool left_open, right_open;
+		start_start = max({this->start_, other.start_});
+		end_end = min({this->end_, other.end_});
+		m = min({start_start, end_end});
+		if ((eq(*end_end, *start_start) and eq(*end_end, *m) and
+			((eq(*end_end, *this->end_) and this->right_open_) or
+			 (eq(*end_end, *other.end_) and other.right_open_))) or
+			(eq(*end_end, *m) and not eq (*end_end, *start_start))) {
+			throw std::runtime_error("not implemented");
+		}
+		else {
+			if (eq(*min({this->start_, other.start_}), *this->start_))
+				start = this->start_;
+			else
+				start = other.start_;
+			if (eq(*max({this->end_, other.end_}), *this->end_))
+				end = this->end_;
+			else
+				end = other.end_;
+			left_open = ((neq(*this->start_, *start) or this->left_open_) and
+							 (neq(*other.start_, *start) or other.left_open_));
+			right_open = ((neq(*this->end_, *end) or this->right_open_) and
+							(neq(*other.end_, *end) or other.right_open_));
+			return interval(start, end, left_open, right_open);
+		}
+	}
+
+	bool Interval::is_subset(const Set &o) const {
+		if (is_a<Interval>(o)) {
+			const Interval &other = static_cast<const Interval &>(o);
+			return this->__eq__(*this->_intersection(other));
+		}
+		else {
+			return o.is_superset(*this);
+		}
+	}
+
+	bool Interval::is_proper_subset(const Set &o) const {
+		if (is_a<Interval>(o)) {
+			const Interval &other = static_cast<const Interval &>(o);
+			return this->__eq__(*this->_intersection(other)) and (not this->__eq__(other));
+		}
+		else {
+			return o.is_proper_superset(*this);
+		}
+	}
+
+	bool Interval::is_superset(const Set &other) const {
+		return other.is_subset(*this);
+	}
+
+	bool Interval::is_proper_superset(const Set &other) const {
+		return other.is_subset(*this) and (not __eq__(other));
+	}
+
+	RCP<const Set> EmptySet::_intersection(const Set &o) const {
+		return emptyset();
+	}
+
+	RCP<const Set> EmptySet::_union(const Set &o) const {
+		if (is_a<Interval>(o)) {
+			const Interval &other = static_cast<const Interval &>(o);
+			return interval(other.start_, other.end_, other.left_open_, other.right_open_);
```

Comment:
If you are returning the argument, then you can take in `const RCP<const Set> &o` and just return it, as @isuruf said.